stan/math/opencl/rev/log1p.hpp

```diff
@@ -0,0 +1,33 @@
+#ifndef STAN_MATH_OPENCL_REV_LOG1P_HPP
+#define STAN_MATH_OPENCL_REV_LOG1P_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the elementwise `log1p()` of a var_value<matrix_cl<double>>.
+ *
+ * @param A argument
+ * @return Elementwise `log1p()` of the input.
+ */
+inline var_value<matrix_cl<double>> log1p(
+    const var_value<matrix_cl<double>>& A) {
+  var_value<matrix_cl<double>> res = log1p(A.val());
+
+  reverse_pass_callback([A, res]() mutable {
+    A.adj() = A.adj() + elt_divide(res.adj(), 1 + A.val());
```

Comment:
```suggestion
    A.adj() = A.adj() + elt_divide(res.adj(), 1.0 + A.val());
```
For kernel generator expressions we can't rely on the compiler to promote constants to appropriate type. So the type conversion would happen in the kernel.